### PR TITLE
Update data loader (PID-to-IDs map + visualisation)

### PR DESCRIPTION
**Pull request by TairYerniyazov**
Requirements: 

* AllClusters_1M.dat has to be placed in data/Herwig/raw.
* data/Herwig/pids_to_ix.pkl needs to be removed to allow HerwigEventMultiHadronDataset to produce a new PID-to-ID map.

### CODE REVIEW

File: hadml/datamodules/components/herwig.py [hunk 1]
```diff
@@ -661,113 +660,296 @@ def get_angles(four_vector):
...
+                data_list = [self._create_data(line, create_dict=True, n_line=i) 
+                             for i, line in enumerate(f) if line.strip()]
+            print()
+            data_list = np.concatenate(data_list)
+            all_data.append(data_list)
+        
+        all_data = np.array(all_data).flatten()
+        count = Counter(all_data)
+        hadron_pids = list(map(lambda x: x[0], count.most_common()))
+        self.pids_to_ix = {pids: i for i, pids in enumerate(hadron_pids)}
```
Comment (remilvus) [inline hadml/datamodules/components/herwig.py:793]: This name actually contains a typo which we never fixed for some reason - should be `pids_to_idx`

File: hadml/datamodules/components/herwig.py [hunk 2]
```diff
@@ -661,113 +660,296 @@ def get_angles(four_vector):
...
+    def _create_data(self, line, cluster_rest_frame=None, create_dict=False, n_line=None):
+        """ Parse data presented as an event prepared in a specific format.
+        May also be used to parse and return only hadron PIDs needed to create a dictionary. """
+
+        # Splitting clusters   
         items = line.split("|")[:-1]
         clusters = [pd.Series(c.split(";")[:-1]).str.split(',', expand=True) for c in items]
 
+        # Creating a dictionary used for mapping particle IDs to indices
+        if create_dict:
```
Comment (remilvus) [inline hadml/datamodules/components/herwig.py:847]: This section can be moved to `_create_dict` for better readability. Only 2 lines of `_create_data` are shared by code executed when `create_dict==True` and `create_dict==False`

File: hadml/datamodules/components/herwig.py [hunk 3]
```diff
@@ -661,113 +660,296 @@ def get_angles(four_vector):
...
+            data, slices = self.collate(all_data_chosen_rest_frame)
+        else:
+            data, slices = self.collate(all_data)
+        torch.save((data, slices), self.processed_paths[0])
 
-            all_data += data_list
 
-        data, slices = self.collate(all_data)
-        torch.save((data, slices), self.processed_paths[0])
+    def _create_dict(self):
```
[the remark this answers is not in the document]
Comment (TairYerniyazov) [reply hadml/datamodules/components/herwig.py:777]: Okay, I'll change it, thanks. At the moment I am reading PyTorch docs as it is likely we will replace the parent class with something simpler than the geometric InMemoryDataset.

File: hadml/datamodules/components/herwig.py (hunk 1 above)
Comment (TairYerniyazov) [reply hadml/datamodules/components/herwig.py:793]: Well, that was a mystery for me all the time but I thought the abbreviation had some sort of explanation. Anyway, I got it, so I will correct it as well.

Comment (remilvus) [reply hadml/datamodules/components/herwig.py:793]: thanks

File: hadml/datamodules/components/herwig.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (remilvus) [reply hadml/datamodules/components/herwig.py:777]: Great!

File: hadml/datamodules/components/herwig.py (hunk 2 above)
Comment (TairYerniyazov) [reply hadml/datamodules/components/herwig.py:847]: Good idea, thanks.

File: configs/datamodule/herwig_event_multihadron.yaml [hunk 4]
```diff
@@ -1,19 +1,14 @@
...
+batch_size: 32
+num_workers: 8
+train_val_test_split: [0.7, 0.15, 0.15]
+
+# Raw file paths (the GAN datamodule will pass those arguments to an appropriate parser)
+raw_file_list:
+  - "AllClusters_10K.dat"
+processed_filename: "herwig_multihadron_events_10K.npy"
+
+# Use it if you want to see invoke data preparation without initialising your PyTorch Lightning model
```
Comment (remilvus) [inline configs/datamodule/herwig_event_multihadron.yaml:13]: Typo (`see invoke`)

File: hadml/datamodules/gan_datamodule.py [hunk 5]
```diff
@@ -251,3 +249,167 @@ def test_dataloader(self):
...
+            raw_file_list=raw_file_list,
+            processed_filename=processed_filename,
+            pid_map_file=pid_map_file,
+            debug=debug
+        )
+        parser.parse_data()
+
+        print('\n', ' '*20, self.__class__.__name__,'\n', '-'*70, sep='')
+
+        if initialise_data_preparation:
```
Comment (remilvus) [inline hadml/datamodules/gan_datamodule.py:288]: we could rename this to something more descriptive, e.g. `log_batch_info`

File: hadml/datamodules/gan_datamodule.py [hunk 6]
```diff
@@ -251,3 +249,167 @@ def test_dataloader(self):
...
+        
+        cluster_kin = data.item()["cluster_kin"]
+        hadron_kin = data.item()["had_kin"]
+        had_type_indices = data.item()["had_type_indices"]
+        cluster_labels = data.item()["cluster_labels"]
+        n_events = len(cluster_kin)
+
+        # Tokenisation
+        generator_cond_input = []
+        discriminator_input = []
```
Comment (remilvus) [inline hadml/datamodules/gan_datamodule.py:312]: For better clarity, these two values can be initialised directly with a helper function containing the two nested `for` loops from below.

File: hadml/datamodules/gan_datamodule.py (hunk 5 above)
Comment (TairYerniyazov) [reply hadml/datamodules/gan_datamodule.py:288]: The "initialise_data_preparation" flag makes an instance call its self.prepared_data() and self.setup() methods even when a MultiHadronEventGANDataModule object is not subsequently passed to any model class. Information about the batch size, as well as the first sample sentences, is printed but it is a collateral action. Therefore renaming it may shift our collaborators' attention towards an action which in fact plays an auxiliary role.

File: hadml/datamodules/gan_datamodule.py (hunk 6 above)
Comment (TairYerniyazov) [reply hadml/datamodules/gan_datamodule.py:312]: Absolutely, thanks. I will change that,

File: configs/datamodule/herwig_event_multihadron.yaml (hunk 4 above)
Comment (TairYerniyazov) [reply configs/datamodule/herwig_event_multihadron.yaml:13]: Thanks.

File: configs/experiment/herwig_event_multihadron.yaml [hunk 7]
```diff
@@ -67,13 +58,3 @@ model:
 
   num_critics: 1
   num_gen: 5
```
Comment (remilvus) [inline configs/experiment/herwig_event_multihadron.yaml:60]: Note for when we start tranining: in the previous paper, we have made one generator step for each critic step.

File: hadml/datamodules/components/herwig.py [hunk 8]
```diff
@@ -25,8 +25,8 @@
 
 from torch_geometric.data import Data
 from torch_geometric.data import InMemoryDataset
-
 from torch.utils.data import Dataset as TorchDataset
+from torch.utils.data import Dataset
 
 pid_map_fname = "pids_to_ix.pkl"
```
Comment (remilvus) [inline hadml/datamodules/components/herwig.py:31]: This is old code, but now with the fixed typo it needs a fix too ("pids_to_idx.pkl")

File: hadml/datamodules/components/herwig.py [hunk 9]
```diff
@@ -660,114 +660,54 @@ def get_angles(four_vector):
...
+        "cluster/hadron padding tokens". It returns a pair of prepared sentences. """
+
+        # Preparing the input sentence. Tokens then need to be concatenated with noise.
+        # [[cluster_kin][cluster_kin]...[padding_token]] 
+        # N = max number of hadrons produced by the heaviest cluster
+        n_hadrons = len(self.hadrons_with_types[index][0])
+        gen_input = torch.cat([self.clusters[index].unsqueeze(0) for _ in range(n_hadrons)])
+        if self.max_n_hadrons - n_hadrons > 0:
+            padding_tokens = torch.cat([self.cluster_padding_token for 
+                                        _ in range(self.max_n_hadrons - n_hadrons)])
```
Comment (remilvus) [inline hadml/datamodules/components/herwig.py:698]: Generator input shouldn't be padded. Intuition: after training when using the model, we won't know how many padding tokens to add

File: hadml/datamodules/components/herwig.py [hunk 10]
```diff
@@ -660,114 +660,54 @@ def get_angles(four_vector):
...
-        q_momenta = np.stack([q_phi, q_theta], axis=1)
-        cond_info = np.concatenate([cs[:, [1,2,3,4]], q1s[:, :1], q2s[:, :1], q_momenta], axis=1)
-        cond_info = torch.from_numpy(cond_info.astype(np.float32))
+        self.n_had_types = n_had_types
+        self.hadrons_with_types = hadrons_with_types
+        self.clusters = clusters
+        self.n_clusters = len(clusters)
+        self.max_n_hadrons = max_n_hadrons
+        # Cluster padding token (zeros)
+        self.cluster_padding_token = torch.zeros(1, len(clusters[0]))
```
Comment (remilvus) [inline hadml/datamodules/components/herwig.py:677]: Not needed (see later comment)

File: hadml/datamodules/gan_datamodule.py [hunk 11]
```diff
@@ -251,3 +250,141 @@ def test_dataloader(self):
...
+            initialise_data_preparation=False,
+            debug=True
+        ):
+        super().__init__()
+        self.processed_filename = os.path.join(os.path.normpath(data_dir),
+                                               "processed", processed_filename)
+        self.train_val_test_split = train_val_test_split
+        self.batch_size = batch_size
+        self.num_workers = num_workers
+        self.data_train: Optional[torch.utils.data.Dataset] = None
```
Comment (remilvus) [inline hadml/datamodules/gan_datamodule.py:274]: It's not important to fix, but just so you know: Typically the dataset is imported directly - `from torch.utils.data import Dataset`. As it is now, it takes up a lot of characters whenever used

Comment (remilvus) [reply hadml/datamodules/gan_datamodule.py:274]: In general, I think most of torch classes are typically either on the top-level (`torch.sth`), or a submodule/class is imported from torch before use.

File: hadml/datamodules/gan_datamodule.py (hunk 5 above)
Comment (remilvus) [reply hadml/datamodules/gan_datamodule.py:289]: Ok, thanks for explaining

File: hadml/datamodules/gan_datamodule.py [hunk 12]
```diff
@@ -251,3 +250,141 @@ def test_dataloader(self):
...
+        # Load data prepared by the parser
+        with open(self.processed_filename, "rb") as f:
+            data = np.load(f, allow_pickle=True)
+        
+        cluster_kin = data.item()["cluster_kin"]
+        hadron_kin = data.item()["had_kin"]
+        had_type_indices = data.item()["had_type_indices"]
+        cluster_labels = data.item()["cluster_labels"]
+        n_events = len(cluster_kin)
+        self.n_had_types = data.item()["n_had_type_indices"] + 1 # 1 extra type for a stop token
```
Comment (remilvus) [inline hadml/datamodules/gan_datamodule.py:310]: Right now it's a padding token. I would change the comment to "1 extra type for a stop/padding token"

File: hadml/datamodules/gan_datamodule.py [hunk 13]
```diff
@@ -251,3 +250,141 @@ def test_dataloader(self):
...
+ 
+        # Assigning hadrons to clusters 
+        n_clusters_extracted_from_events = 0
+        self.clusters, self.hadrons_with_types = [], []
```
Comment (remilvus) [inline hadml/datamodules/gan_datamodule.py:314]: In the comment from before, I meant that we should refactor lines 316-333 into a helper function and call it here. So we would have ` self.clusters, self.hadrons_with_types = self._helper(...)`.

This way, understanding what happens in `prepare_data` will be easier

File: hadml/datamodules/gan_datamodule.py [hunk 14]
```diff
@@ -251,3 +250,141 @@ def test_dataloader(self):
...
+
+        for i in range(n_events):
+            cluster_in_event = []  # [cluster ... cluster]
+            hadrons_in_event = []  # [ [[hadron ... hadron][h_id ... h_id]] ... [[...][...]] ]
```
Comment (remilvus) [inline hadml/datamodules/gan_datamodule.py:318]: I'm not saying we have to change something here, but let's discuss whether we want to use dataclasses etc. at one of our meetings

**Review by remilvus — CHANGES_REQUESTED** [review]

File: hadml/datamodules/components/herwig.py [hunk 15]
```diff
@@ -660,114 +660,48 @@ def get_angles(four_vector):
...
+
+    def __getitem__(self, index):
+        """ This method is mainly responsible for tokenisation and building sentences containing 
+        "cluster/hadron padding tokens". It returns a pair of prepared sentences. """
+
+        # Preparing the input sentence. Tokens then need to be concatenated with noise.
+        # [[cluster_kin][cluster_kin]...[padding_token]] 
+        # N = max number of hadrons produced by the heaviest cluster
+        n_hadrons = len(self.hadrons_with_types[index][0])
+        gen_input = torch.cat([self.clusters[index].unsqueeze(0) for _ in range(self.max_n_hadrons)])
```
Comment (remilvus) [inline hadml/datamodules/components/herwig.py:693]: To simplify this line, you can do `torch.stack` instead of combining `stack` and `unsqueeze`

File: hadml/datamodules/components/herwig.py [hunk 16]
```diff
@@ -660,114 +660,48 @@ def get_angles(four_vector):
...
+
+
+    def __getitem__(self, index):
+        """ This method is mainly responsible for tokenisation and building sentences containing 
+        "cluster/hadron padding tokens". It returns a pair of prepared sentences. """
+
+        # Preparing the input sentence. Tokens then need to be concatenated with noise.
+        # [[cluster_kin][cluster_kin]...[padding_token]] 
+        # N = max number of hadrons produced by the heaviest cluster
+        n_hadrons = len(self.hadrons_with_types[index][0])
```
Comment (remilvus) [inline hadml/datamodules/components/herwig.py:692]: For clarity, `n_hadrons` can be initialised closer to its usage

File: hadml/datamodules/components/herwig.py [hunk 17]
```diff
@@ -660,114 +660,48 @@ def get_angles(four_vector):
...
+        return self.n_clusters
+
+
+    def __getitem__(self, index):
+        """ This method is mainly responsible for tokenisation and building sentences containing 
+        "cluster/hadron padding tokens". It returns a pair of prepared sentences. """
+
+        # Preparing the input sentence. Tokens then need to be concatenated with noise.
+        # [[cluster_kin][cluster_kin]...[padding_token]] 
+        # N = max number of hadrons produced by the heaviest cluster
```
Comment (remilvus) [inline hadml/datamodules/components/herwig.py:691]: Is `N` in the comment the same as `n_hadrons`? Using the same variable would be clearer

File: hadml/datamodules/components/herwig.py [hunk 18]
```diff
@@ -660,114 +660,48 @@ def get_angles(four_vector):
...
+        n_hadrons = len(self.hadrons_with_types[index][0])
+        gen_input = torch.cat([self.clusters[index].unsqueeze(0) for _ in range(self.max_n_hadrons)])
+        
+        # Preparing the output sentence. The hadron type is a one-hot vector.
+        # [[hadron_kin, hadron_type][hadron_kin, hadron_type]...[padding_token]]
+        # N = max number of hadrons produced by the heaviest cluster
+        hadrons = self.hadrons_with_types[index][0]
+        hadron_types = self.hadrons_with_types[index][1]
+        hadron_types_ohe = torch.nn.functional.one_hot(hadron_types, self.n_had_types)
```
Comment (remilvus) [inline hadml/datamodules/components/herwig.py:700]: We can also consider doing ` hadron_types = torch.nn.functional.one_hot(hadron_types, self.n_had_types)` or ` hadron_types = torch.nn.functional.one_hot(self.hadrons_with_types[index][1], self.n_had_types)`

File: hadml/datamodules/gan_datamodule.py [hunk 19]
```diff
@@ -251,3 +251,196 @@ def test_dataloader(self):
...
+
+        self.max_n_hadrons = max([len(hadron_with_type[0]) for hadron_with_type in 
+                                  self.hadrons_with_types])
+        
+        print("Initial number of events:", n_events)
+        print("Total number of clusters:", n_clusters_extracted_from_events)
+        print("Total number of hadron types (with a stop token):", self.n_had_types)
+        print("Largest number of hadrons per cluster:", self.max_n_hadrons)
+
+    def __get_hadrons_and_clusters__(self, n_events, cluster_kin, cluster_labels, hadron_kin,
```
Comment (remilvus) [inline hadml/datamodules/gan_datamodule.py:342]: Double leading/trailing underscores are generally only used for python "magic" methods. Here, one leading underscore should be enough: `_get_hadrons_and_clusters`

see: https://peps.python.org/pep-0008/#method-names-and-instance-variables

File: hadml/metrics/compare_fn.py [hunk 20]
```diff
@@ -263,3 +262,15 @@ def __call__(
         plt.close("all")
 
         return out_images
+
+
+class CompareMultiHadronEventGan():
```
Comment (remilvus) [inline hadml/metrics/compare_fn.py:267]: Although we did it like this before, there is no need for `CompareMultiHadronEventGan` to be a separate class. Its very unlikely we will be using more than one plotting function, so we don't need the modularity. This just makes the code needlessly complicated.

We can draw the plots using a method of `MultiHadronEventGANModule`

**Comment by TairYerniyazov** [comment]
Changes have been withdrawn (will be visible in the next PR) as asked, so this PR related to visualisation is ready.
